packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1159 @@
+  /// It is also the primary respondent for the [backButtonDispatcher]. The
+  /// [Router] relies on the [RouterDelegate.popRoute] to handles the back
+  /// button intends.
+  ///
+  /// If the [RouterDelegate.currentConfiguration] returns a non-null object,
+  /// this [Router] will opt for URL updates.
+  final RouterDelegate<T> routerDelegate;
+
+  /// Retrieve the router of exact type.
+  static Router<R> of<R>(BuildContext context) {
+    final _RouterScope scope = context.dependOnInheritedWidgetOfExactType<_RouterScope>();
+    assert(scope != null);
+    return scope.routerState.widget as Router<R>;
+  }
+
+  /// Forces the [Router] to to run the [callback] with the assumption whether
+  /// it is a navigating event.
+  ///
+  /// Normally, calls to this method is not necessary because the [Router] will
+  /// detect whether the callback is a navigating event automatically as soon
+  /// as it is run. Only use this method when you want to force the [Router]
+  /// to recognize or ignore the [callback] as navigation event.
+  ///
+  /// When [router] detect a navigating event, it will report that event to the
+  /// [routeInformationProvider]. This is useful in the web application because
+  /// the web platform relies on these reporting to support browser backward and
+  /// forward buttons.
+  ///
+  /// if the [isNavigationEvent] is true, the [Router] will treat the [callback]
+  /// as a navigating event regardless and report the event to the
+  /// [routeInformationProvider].
+  ///
+  /// if the [isNavigationEvent] is false, the [Router] will treat the [callback]
+  /// as a non-navigating event regardless and will not report to the
+  /// [routeInformationProvider].
+  ///
+  /// if nothing is passed into [isNavigatingEvent], the [Router] will
+  /// decide whether the [callback] is a navigating event by itself.
+  static void run(
+    BuildContext context,
+    VoidCallback callback, {
+    bool isNavigatingEvent,
+  }) {
+    final _RouterScope scope = context.getElementForInheritedWidgetOfExactType<_RouterScope>().widget as _RouterScope;
+
+    if (isNavigatingEvent == true)
+      scope.routerState._reportingStatus = _RouteInformationReportingStatus.must;
+    if (isNavigatingEvent == false)
+      scope.routerState._reportingStatus = _RouteInformationReportingStatus.ignore;
+    scope.routerState.setState(callback);
+  }
+
+  @override
+  StatefulElement createElement() => _RouterStatefulElement(this);
+
+  @override
+  State<Router<T>> createState() => _RouterState<T>();
+}
+
+typedef _AsyncPassthrough<Q> = Future<Q> Function(Q);
+
+class _RouterStatefulElement extends StatefulElement {
+  _RouterStatefulElement(StatefulWidget widget) : super(widget);
+
+  _RouterState<dynamic> get routerState => state as _RouterState<dynamic>;
+
+  @override
+  void rebuild() {
+    super.rebuild();
+    routerState._reportRouteInformation();
+  }
+}
+
+enum _RouteInformationReportingStatus {
+  none,
+  maybe,
+  must,
+  ignore,
+}
```

Comment:
It would be helpful to have some implementation comments explaining what these mean.